### PR TITLE
✨ [RUMF-1510] Warn the user when a heavy context is used

### PR DESCRIPTION
**Pull request by amortemousque**
## Motivation

Raise awareness about customer data impact on the user bandwidth.

## Changes
- For ssr compat: 
  - Move `getGlobalObject` on its own file to avoid circular ref with `timer.ts`
  - Use `getGlobalObject` in `setTimeout` and `setInterval`
- Compute customer data bytes count at API call
- Add a console warning when 3KiB customer data reach for
    - Global context 
    - User
    - Feature flag evaluation
    - Logger context (I made the call since it also uses context manager)

## Testing

- [x] Local
- [ ] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/core/src/tools/heavyCustomerDataWarning.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { display } from './display'
+import { ONE_KIBI_BYTE } from './utils'
+
+export const CUSTOMER_DATA_BYTES_LIMIT = 3 * ONE_KIBI_BYTE
```
Comment (bcaudan) [inline packages/core/src/tools/heavyCustomerDataWarning.ts:4]: **💬 suggestion:** ‏could you add a comment with a bit of reasoning on the value?

File: packages/core/src/tools/contextManager.ts [hunk 2]
```diff
@@ -1,60 +1,65 @@
...
-export function createContextManager(computeBytesCountImpl = computeBytesCount) {
+export function createContextManager(customerDataType: CustomerDataType, computeBytesCountImpl = computeBytesCount) {
   let context: Context = {}
-  let bytesCountCache: number | undefined
+  let bytesCountCache: number
+
+  const { throttled: computeBytesCountThrottled } = throttle((context: Context) => {
+    bytesCountCache = computeBytesCountImpl(jsonStringify(context)!)
+    warnIfCustomerDataLimitReached(bytesCountCache, customerDataType)
+  }, BYTES_COMPUTATION_THROTTLING_DELAY)
```
Comment (bcaudan) [inline packages/core/src/tools/contextManager.ts:18]: **💬 suggestion:** ‏Could you add a bit of reasoning and/or a test for this throttle use?
(same with feature flag context)

File: packages/core/src/tools/contextManager.ts [hunk 3]
```diff
@@ -1,60 +1,65 @@
-import { computeBytesCount, deepClone, jsonStringify } from './utils'
+import { computeBytesCount, deepClone, jsonStringify, ONE_KIBI_BYTE, throttle } from './utils'
 import type { Context, ContextValue } from './context'
+import type { CustomerDataType } from './heavyCustomerDataWarning'
+import { warnIfCustomerDataLimitReached } from './heavyCustomerDataWarning'
+
+export const CUSTOMER_CONTEXT_BYTES_LIMIT = 5 + ONE_KIBI_BYTE
```
[the remark this answers is not in the document]
Comment (amortemousque) [reply packages/core/src/tools/contextManager.ts]: erf..

File: packages/core/src/tools/heavyCustomerDataWarning.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { display } from './display'
+import { ONE_KIBI_BYTE } from './utils'
+
+// RUM batch bytes limit is 16KB
+// ensure that we leave room for other event attributes and maintain a decent amount of event per batch
+// 3KB (customer data) + 1KB (other attributes) * 4 (events per batch) = 16KB
```
Comment (bcaudan) [inline packages/core/src/tools/heavyCustomerDataWarning.ts:6]: **🥜 nitpick:** ‏operator priorities
```suggestion
// (3KB (customer data) + 1KB (other attributes)) * 4 (events per batch) = 16KB
```

File: packages/core/src/tools/heavyCustomerDataWarning.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { display } from './display'
+import { ONE_KIBI_BYTE } from './utils'
+
+// RUM batch bytes limit is 16KB
```
Comment (bcaudan) [inline packages/core/src/tools/heavyCustomerDataWarning.ts:4]: **🥜 nitpick:**
```suggestion
// RUM and logs batch bytes limit is 16KB
```

**Review by BenoitZugmeyer — APPROVED** [review]

**Pull request merged**